### ISSUE TITLE
Parallel DELETE /v2/apps/guid on same app result in 204/500 instead of 204/404

### ISSUE THREAD

**Issue by stephanme** [cloudfoundry/cloud_controller_ng#1124]
## Issue

If there are two DELETE requests on the same app running in parallel then one request succeeds and the other fails. This is expected.

However, the failing request results in a 500 response (internal server error) instead of a 404 response (not found). 

This renders statistics about failed CC api calls useless because I can't distinguish between CF errors and user errors.

Also users open bugs instead of checking their 'interesting' CF api usage.

## Context

cf-deployment v1.21.0
capi-release v1.52.0

## Steps to Reproduce

```
$ cf push -n d047883test
Pushing from manifest to org uptime / space uptimeci as d047883...
Using manifest file /home/stephan/d047883/git/staticfile-sample/manifest.yml
...
Waiting for app to start...

name:              staticfile-test
requested state:   started
instances:         1/1
usage:             64M x 1 instances
routes:            d047883test.cfapps.sap.hana.ondemand.com
last uploaded:     Tue 10 Apr 17:08:56 DST 2018
stack:             cflinuxfs2
buildpack:         staticfile
start command:     $HOME/boot.sh

     state     since                  cpu    memory     disk      details
#0   running   2018-04-10T15:09:10Z   0.0%   0 of 64M   0 of 1G

$ CF_TRACE=delete1.log cf d staticfile-test -f &  CF_TRACE=delete2.log cf d staticfile-test -f &
[1] 10274
[2] 10275
$ Deleting app staticfile-test in org uptime / space uptimeci as d047883...
Deleting app staticfile-test in org uptime / space uptimeci as d047883...
FAILED
Server error, status code: 500, error code: 10001, message: An unknown error occurred.
OK

[1]-  Done                    CF_TRACE=delete1.log cf d staticfile-test -f
[2]+  Exit 1                  CF_TRACE=delete2.log cf d staticfile-test -f
$
```
## Expected result

The failed cf delete fails with a 404.

This actually happens for a single DELETE via cf curl on a non-existing application:

```
$ cf curl "/v2/apps/53897c20-af3c-4121-b52f-ae93e6cff33a?async=true&recursive=true" -v -i -X DELETE

REQUEST: [2018-04-10T17:27:21+02:00]
DELETE /v2/apps/53897c20-af3c-4121-b52f-ae93e6cff33a?async=true&recursive=true HTTP/1.1
Host: api.cf.sap.hana.ondemand.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.34.1+bbdf81482.2018-01-17 / linux



RESPONSE: [2018-04-10T17:27:21+02:00]
HTTP/1.1 404 Not Found
Connection: close
Content-Length: 140
Content-Type: application/json;charset=utf-8
Date: Tue, 10 Apr 2018 15:27:21 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Ratelimit-Limit: 80000
X-Ratelimit-Remaining: 79872
X-Ratelimit-Reset: 1523375144
X-Vcap-Request-Id: abf6eac0-821b-4e1d-4140-c2393069d615::7899aa14-9b2b-490c-bc77-9d44dad52c53

{
  "description": "The app could not be found: 53897c20-af3c-4121-b52f-ae93e6cff33a",
  "error_code": "CF-AppNotFound",
  "code": 100004
}
```

## Current result

The failed cf delete fails with a 500.

See attached CF_TRACE log files:
- delete1.log - cf delete succeeded
- delete2.log - cf delete failed with 500 instead of 404

**Comment by stephanme** [comment]
[delete1.log](https://github.com/cloudfoundry/cloud_controller_ng/files/1895316/delete1.log)
[delete2.log](https://github.com/cloudfoundry/cloud_controller_ng/files/1895317/delete2.log)

**Comment by stephanme** [comment]
I found the following stacktrace in the CC logs for the cf delete that responded 500:

```
{"timestamp":1523265948.8812506,"message":"Request failed: 500: {\"error_code\"=>\"UnknownError\", \"description\"=>\"An unknown error occurred.\", \"code\"=>10001, \"test_mode_info\"=>{\"description\"=>\"Record not found\", \"error_code\"=>\"CF-NoExistingObject\", \"backtrace\"=>[
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/model/base.rb:2174:in `_refresh'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/model/base.rb:1670:in `lock!'\",
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/app_delete.rb:22:in `block (2 levels) in delete'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:224:in `_transaction'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:199:in `block in transaction'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `block in synchronize'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/connection_pool/threaded.rb:107:in `hold'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/connecting.rb:301:in `synchronize'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sequel-4.49.0/lib/sequel/database/transactions.rb:165:in `transaction'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/app_delete.rb:21:in `block in delete'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/app_delete.rb:20:in `each'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/app_delete.rb:20:in `delete'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/app_delete.rb:36:in `delete_without_event'\", 
\"/var/vcap/data/packages/cloud_controller_ng/0dcd5609db985649c15758056306ce1bf6026e30/cloud_controller_ng/app/controllers/runtime/apps_controller.rb:142:in `delete'\", 
\"/var/vcap/data/packages/cloud_controller_ng/0dcd5609db985649c15758056306ce1bf6026e30/cloud_controller_ng/app/controllers/base/base_controller.rb:79:in `dispatch'\", 
\"/var/vcap/data/packages/cloud_controller_ng/0dcd5609db985649c15758056306ce1bf6026e30/cloud_controller_ng/lib/cloud_controller/rest_controller/routes.rb:16:in `block in define_route'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1610:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1610:in `block in compile!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:975:in `block (3 levels) in route!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:994:in `route_eval'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:975:in `block (2 levels) in route!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1015:in `block in process_route'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1013:in `catch'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1013:in `process_route'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:973:in `block in route!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:972:in `each'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:972:in `route!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1085:in `block in dispatch!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1067:in `block in invoke'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1067:in `catch'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1067:in `invoke'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1082:in `dispatch!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:907:in `block in call!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1067:in `block in invoke'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1067:in `catch'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:1067:in `invoke'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:907:in `call!'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:895:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-protection-1.5.3/lib/rack/protection/xss_header.rb:18:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-protection-1.5.3/lib/rack/protection/path_traversal.rb:16:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-protection-1.5.3/lib/rack/protection/json_csrf.rb:18:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-protection-1.5.3/lib/rack/protection/base.rb:49:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-protection-1.5.3/lib/rack/protection/base.rb:49:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-protection-1.5.3/lib/rack/protection/frame_options.rb:31:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-1.6.8/lib/rack/nulllogger.rb:9:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-1.6.8/lib/rack/head.rb:13:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:182:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/sinatra-1.4.8/lib/sinatra/base.rb:2013:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-1.6.8/lib/rack/urlmap.rb:66:in `block in call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-1.6.8/lib/rack/urlmap.rb:50:in `each'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-1.6.8/lib/rack/urlmap.rb:50:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/rate_limiter.rb:43:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_logs.rb:22:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/security_context_setter.rb:19:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/vcap_request_id.rb:15:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/cors.rb:49:in `call_app'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/cors.rb:14:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/middleware/request_metrics.rb:12:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/rack-1.6.8/lib/rack/builder.rb:153:in `call'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/thin-1.7.0/lib/thin/connection.rb:86:in `block in pre_process'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/thin-1.7.0/lib/thin/connection.rb:84:in `catch'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/thin-1.7.0/lib/thin/connection.rb:84:in `pre_process'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/thin-1.7.0/lib/thin/connection.rb:50:in `block in process'\", 
\"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.4.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:1067:in `block in spawn_threadpool'\"]}}","log_level":"error","source":"cc.api","data":{"request_guid":"0e16c743-cc11-4379-5f35-86310bad85df::c97bb52c-d1fb-46a1-ab82-c72876863a81"},"thread_id":47089514721240,"fiber_id":47089497083260,"process_id":5170,"file":"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/lib/sinatra/vcap.rb","lineno":46,"method":"block in registered"}
```

**Comment by selzoc** [comment]
Hi @stephanme, thanks for reporting this!

We have fixed this in 637c255664930b9df09683805c8e50dc943ffcfb and it will most likely appear in the next capi release.

Thanks,
Chris && @ericpromislow

**Comment by zrob** [comment]
Acctually fixed in  51e185f
